text/3334-niche.md

```diff
@@ -0,0 +1,348 @@
+
+The attribute `#[niche]` may only appear on a struct declaration. The struct
+must contain exactly one field of a non-zero-sized type (non-ZST). The struct
+may contain zero or more ZST fields, such as `PhantomData`. (Note that
+`#[non_exhaustive]` types do not count as ZSTs for this purpose, even if they
+*currently* contain no fields with non-zero sizes.)
+
+Declaring a niche on any item other than a struct declaration results in an
+error.
+
+Declaring a niche on a struct containing more or less than one non-zero-sized
+field results in an error.
+
+Declaring multiple `niche` attributes on a single item, or multiple key-value
+pairs within a single `niche` attribute, results in an error.
+
+Declaring a niche on a struct that has any generic parameters affecting the
+non-zero-sized field results in an error.
+
+Declaring a range niche with an empty range (e.g. `0..0`) results in a
+warn-by-default lint. As with many lints, this lint should be automatically
+suppressed for code expanded from a macro.
+
+Declaring a range niche with an invalid range (e.g. `5..0`) results in an
+error.
+
+Declaring a niche using a negative value or a negative range endpoint results
+in an error. The representation of negative values depends on the size of the
+type, and the compiler may not have that information at the time it handles
+attributes such as `niche`. The text of the error should suggest the
+appropriate two's-complement unsigned equivalent to use. The compiler may
+support this in the future.
+
+Declaring a range niche with an open start (`..3`) results in an error, for
+forwards-compatibility with support for negative values.
+
+Declaring a niche using a non-literal value (e.g. `usize::MAX`) results in an
+error. Constants can use compile-time evaluation, and compile-time evaluation
+does not occur early enough for attributes such as niche declarations.
+
+If a type `T` contains multiple niche values (e.g. `#[niche(range = 8..16)]`),
+the compiler does not guarantee any particular usage of those niche values in
+the representation of types containing `T`. In particular, the
+compiler does not commit to making use of all the invalid values of the niche,
+even if it otherwise could have.
+
+However, multiple instances of the same identical type (e.g. `Option<T>` and
+`Option<T>`) will use an identical representation (whether the type contains a
+niche or not). This permits a round-trip between such a value and a byte
+representation.
+
+If a type `T` contains niches and uses `repr(C)` or `repr(transparent)`, the
+compiler guarantees to use the same storage size for the type as it would
+without the niche, even if the niche might allow storing fewer bytes. If a type
+`T` contains niches and uses the default (`Rust`) `repr`, the compiler may
+choose to represent the type using fewer bytes if the niche would allow doing
+so. For instance:
+
+```rust
+#[niche(range = 4..)]
+struct S {
+    field: u16,
+}
+
+// `size_of::<S>()` may return less than 2
```

Comment:
While the layout of `repr(Rust)` is of course unspecified, I don't think explicitly proposing this here is a good idea. Giving `S` only a size of 1 will trivially break as soon as anyone takes a `&mut u16` to the field and writes any value to it, as that write will write 2 bytes. So `S` will always have to be at least as big as its fields (unless the compiler can prove that no reference is ever taken to the field, which doesn't seem too useful).

Reply:
By my reading of the RFC this would cause unsoundness in safe code:

```rust
#[niche(range = 4..)]
struct S {
  field: u16,
}

// on a compiler implementing this optimization
assert_eq!(size_of::<S>(), 1);

// since these are 1-byte values, the array is 2 bytes
let values: [S; 2] = [S(0), S(0)];

// mutable references are called out as `unsafe`,
// so I assume non-mutable references to the field are safe to construct
let refs: [&u16; 2] = [&values[0].0, &values[1].0];

// Since the values are one byte apart, one of these two references must be unaligned
dbg!(*refs[0], *refs[1]);
```

Reply:
You're entirely right. I was attempting to allow leeway for Rust to aggressively optimize the structure, but one of the points of a niche is to still allow references (including mutable references) to the field using its normal type. So, in the absence of move types or similar, this doesn't work. I've removed it, and added it to the future possibilities section.